File: BackEnd/Domain/Entities/Factura.cs
namespace SistemaFacturacion.Domain.Entities
{
    public class Factura
    {
        public int IdFactura { get; set; }
        public int NumeroFactura { get; set; }
        public DateTime Fecha { get; set; }
        public int IdCliente { get; set; }
        public int? IdUsuario { get; set; }


        public decimal Subtotal { get; set; }
        public decimal Iva { get; set; }
        public decimal Total { get; set; }
        public string? ClienteNombreHistorico { get; set; }
        public string? ClienteApellidoHistorico { get; set; }
        public string? ClienteCorreoHistorico { get; set; }
        public string? ClienteTelefonoHistorico { get; set; }
        public string? ClienteDireccionHistorico { get; set; }
        public string? VendedorNombreHistorico { get; set; }
        public string? VendedorCorreoHistorico { get; set; }
        public string? VendedorRolHistorico { get; set; }


        public Cliente? Cliente { get; set; }
        public Usuario? Usuario { get; set; }
        public ICollection<DetalleFactura> DetallesFactura { get; set; } = new List<DetalleFactura>();
    }
}

File: FrontEnd/js/core/entities/Factura.js
import { Dinero } from "../valueObjects/Dinero.js";
import { DomainException } from "../exceptions/DomainException.js";
import { DetalleFactura } from "./DetalleFactura.js";

export class Factura {
  constructor({ idFactura = 0, numeroFactura = 0, fecha = new Date(), cliente = null, idUsuario = null }) {
    this.idFactura = idFactura;
    this.numeroFactura = numeroFactura;
    this.fecha = fecha instanceof Date ? fecha : new Date(fecha);
    this.cliente = cliente;
    this.idUsuario = idUsuario;
    this.detallesFactura = [];
    
    this.subtotal = 0;
    this.iva = 0;
    this.total = 0;
  }

  setCliente(cliente) {
    if (!cliente) {
      throw new DomainException("El cliente es obligatorio para la factura.");
    }
    this.cliente = cliente;
  }

  agregarProducto(producto, cantidad) {
    if (!this.cliente) {
      throw new DomainException("Debe seleccionar un cliente antes de agregar productos.");
    }
    if (!producto) {
      throw new DomainException("El producto es obligatorio.");
    }
    producto.validarStockDisponible(cantidad);

    // Si ya existe en el detalle, sumamos la cantidad
    const existenteIdx = this.detallesFactura.findIndex(d => d.idProducto === producto.idProducto);
    if (existenteIdx !== -1) {
      const nuevaCantidad = this.detallesFactura[existenteIdx].cantidad + cantidad;
      producto.validarStockDisponible(nuevaCantidad);
      
      // Reemplazar la línea por una nueva con cantidad acumulada
      this.detallesFactura[existenteIdx] = new DetalleFactura({
        idProducto: producto.idProducto,
        nombreProducto: producto.nombre,
        precioUnitario: producto.precio,
        cantidad: nuevaCantidad,
        aplicaIva: producto.aplicaIva,
        porcentajeIva: producto.porcentajeIva
      });
    } else {
      this.detallesFactura.push(new DetalleFactura({
        idProducto: producto.idProducto,
        nombreProducto: producto.nombre,
        precioUnitario: producto.precio,
        cantidad: cantidad,
        aplicaIva: producto.aplicaIva,
        porcentajeIva: producto.porcentajeIva
      }));
    }

    this.calcularTotales();
  }

  quitarDetalle(idProducto) {
    this.detallesFactura = this.detallesFactura.filter(d => d.idProducto !== idProducto);
    this.calcularTotales();
  }

  calcularTotales() {
    let subtotalAcumulado = 0;
    let ivaAcumulado = 0;

    this.detallesFactura.forEach(d => {
      subtotalAcumulado += d.subtotal;
      ivaAcumulado += d.ivaLinea;
    });

    this.subtotal = Dinero.round(subtotalAcumulado);
    this.iva = Dinero.round(ivaAcumulado);
    this.total = Dinero.round(this.subtotal + this.iva);
  }

  validarParaGuardar() {
    if (!this.cliente) {
      throw new DomainException("Debe seleccionar un cliente.");
    }
    if (this.detallesFactura.length === 0) {
      throw new DomainException("La factura debe tener al menos un producto.");
    }
  }
}
